src/ring_theory/derivations.lean

```diff
@@ -0,0 +1,200 @@
+
+/-- This class register an instance that the scalar multiplication over `M` is indeed defined as
+above in `transitive_scalar`. -/
+@[protect_proj]
+class compatible_semimodule (R : Type*) (A : Type*) [comm_semiring R] [semiring A] [algebra R A]
+(M : Type*) [add_comm_monoid M] [semimodule A M] [semimodule R M] :=
+(compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m)
+
+instance algebra.to_compatible_semimodule  {R : Type*} {A : Type*} [comm_semiring R] [semiring A]
+  [algebra R A] : compatible_semimodule R A A :=
+⟨λ r a, by rw [smul_def]; refl⟩
+
+section compatible_semimodule
+
+variables {R : Type*} [comm_semiring R]
+(A : Type*) [comm_semiring A] [algebra R A]
+{M : Type*} [add_comm_monoid M] [semimodule A M] [semimodule R M] [compatible_semimodule R A M]
+{N : Type*} [add_comm_monoid N] [semimodule A N] [semimodule R N] [compatible_semimodule R A N]
+
+@[simp] lemma compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m :=
+compatible_semimodule.compatible_smul r m
+
+variable {A}
+
+@[simp] lemma compatible_smul_comm (r : R) (a : A) (m : M) : a • r • m = r • a • m :=
+by rw [compatible_smul A r m, compatible_smul A r (a • m), ←mul_smul, mul_comm, mul_smul]
+
+@[simp] lemma compatible_map_smul (f : M →ₗ[A] N) (r : R) (m : M) :
+  f (r • m) = r • f m :=
+by rw [compatible_smul A r m, linear_map.map_smul, ←compatible_smul A r (f m)]
+
+instance : has_coe (M →ₗ[A] N) (M →ₗ[R] N) :=
+⟨λ f, ⟨f.to_fun, λ x y, f.map_add' x y, λ r n, compatible_map_smul _ _ _⟩⟩
+
+end compatible_semimodule
+
+/-- Derivations in commutative algebra. To be changed.-/
+@[protect_proj]
+structure derivation (R : Type*) (A : Type*) [comm_semiring R] [comm_semiring A]
+  [algebra R A] (M : Type*) [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+  [compatible_semimodule R A M]
+  extends A →ₗ[R] M :=
+(leibniz' (a b : A) : to_fun (a * b) = a • to_fun b + b • to_fun a)
+
+namespace derivation
+
+section
+
+variables {R : Type*} [comm_semiring R]
+{A : Type*} [comm_semiring A] [algebra R A]
+{M : Type*} [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+[compatible_semimodule R A M]
+(D : derivation R A M) {D1 D2 : derivation R A M} (r : R) (a b : A)
+
+instance : has_coe_to_fun (derivation R A M) := ⟨_, λ D, D.to_linear_map.to_fun⟩
+
+instance has_coe_to_linear_map : has_coe (derivation R A M) (A →ₗ[R] M) :=
+  ⟨λ D, D.to_linear_map⟩
+
+@[simp, norm_cast]
+lemma coe_linear_map (f : derivation R A M) :
+  ⇑(f : A →ₗ[R] M) = f := rfl
+
+lemma coe_injective (H : ⇑D1 = D2) : D1 = D2 :=
+by cases D1; cases D2; congr'; exact linear_map.coe_inj H
+
+@[ext] theorem ext (H : ∀ a, D1 a = D2 a) : D1 = D2 :=
+coe_injective $ funext H
+
+@[simp] lemma map_add : D (a + b) = D a + D b := is_add_hom.map_add D a b
+@[simp] lemma map_zero : D 0 = 0 := is_add_monoid_hom.map_zero D
+@[simp] lemma leibniz : D (a * b) = a • D b + b • D a := D.leibniz' _ _
+@[simp] lemma map_smul : D (r • a) = r • D a := linear_map.map_smul D r a
```

Comment:
I would either rename `leibniz` to `map_mul`, or change the order so that `leibniz` doesn't break the pattern.
```lean
@[simp] lemma map_add : D (a + b) = D a + D b := is_add_hom.map_add D a b
@[simp] lemma map_zero : D 0 = 0 := is_add_monoid_hom.map_zero D
@[simp] lemma map_smul : D (r • a) = r • D a := linear_map.map_smul D r a

@[simp] lemma leibniz : D (a * b) = a • D b + b • D a := D.leibniz' _ _
```